### ISSUE TITLE
OAuth problem when uploading attatchement to invoice

### ISSUE THREAD

**Issue by NickBarrett** [XeroAPI/Xero-Net#48]
I am getting the following error when trying to attach a pdf document to an invoice.
oauth_problem=signature_invalid
oauth_problem_advice=Failed to validate signature

Here is the code that I am trying to run:
```
Xero.Api.Core.Model.Attachment attachment = new Xero.Api.Core.Model.Attachment(stream, fileName, "application/pdf", fileSize);
_repository.Attachments.AddOrReplace(attachment, Xero.Api.Core.Model.Types.AttachmentEndpointType.Invoices, invoiceAccRef);
```

My program is targeting .Net 4.5 and is using a PrivateAuthenticator in the XeroCoreApi constructor. I have been using the api fine to create invoices and credit notes as well as sync contact and organisation information with my local database.

Apologies if I'm simply not using the api correctly. Any help much appreciated.

**Comment by davidjnet** [comment]
I had this problem when the attachment filename had spaces in it.

I worked around the problem by changing spaces to underscores.

The filename forms part of the query string, which in turn forms part of the OAuth parameters, which in turn gets signed. I suspect there is a bug in how it escapes space characters, hence the "signature_invalid" error.

**Comment by rowley-xero** [comment]
Thanks for the feed back. I'll check that the Dust library we are using is encoding the paramters correctly. Spaces should become %20 in the signaturebase.

**Comment by NickBarrett** [comment]
Thanks for that David. I can confirm that spaces in filenames cause this problem. I did as you said and replaced them with underscores and the upload worked fine.

**Comment by NickBarrett** [comment]
So I've done a bit more digging. From what I can tell, when adding an attachment to an invoice, the file name becomes part of the uri instead of being a parameter. So the uri for the POST ends up looking something like this:

```
https://api.xero.com/api.xro/2.0/Invoices/11111111-2222-3333-4444-555555555555/Attachments/file with spaces in name.pdf
```

The RequestUri that is sent to create the OAuth signature ends up looking like:

```
https://api.xero.com/api.xro/2.0/Invoices/11111111-2222-3333-4444-555555555555/Attachments/file with spaces in name.pdf?unitdp=4&page=1&IncludeOnline=true
```

I'm using the Xero.Api.Example.Applications.Private.PrivateAuthenticator. Inside the RsaSha1Signer, the signatureBaseString that is sent to the Sign method of the RsaSha1 object looks like:

```
POST&https%3A%2F%2Fapi.xero.com%2Fapi.xro%2F2.0%2FInvoices%2F11111111-2222-3333-4444-555555555555%2FAttachments%2Ffile%20with%20spaces%20in%20name.pdf&IncludeOnline%3Dtrue%26oauth_consumer_key...
```

This appears to have the %20's in the correct place. I'm not sure what this means but I thought it may help someone else figure out what's happening.

**Comment by smax48** [comment]
I also had this issue when the file name contains spaces.
Additionally, the AttachmentsEndpoint works incorrectly with Client.Parameters: this collection is **not initialized** and **not cleared** after an http call. This leads to NullObjectReference exception if you try to use this endpoint before anything else. 
Other endpoints inherit from base classes (e.g. XeroReadEndpoint) with proper handling of the Parameters collection, but AttachmentsEndpoint doesn't have a base class.

**Comment by MJMortimer** [comment]
This sneaky issue should finally be resolved in #126
